Core_OverlayMods/TextureSaveHandler.cs

```diff
@@ -0,0 +1,394 @@
+            {
+                if (!Directory.Exists(LocalTexturePath))
+                    Directory.CreateDirectory(LocalTexturePath);
+
+                var dict = new Dictionary<KeyValuePair<int, ulong>, byte[]>();
+                // KoiClothesOverlayX
+                Dictionary<CoordinateType, Dictionary<string, ClothesTexData>> clothesData = null;
+                if (dictRaw is Dictionary<CoordinateType, Dictionary<string, ClothesTexData>> clothesDataReceived)
+                {
+                    clothesData = clothesDataReceived.ToDictionary(
+                        kvp => kvp.Key,
+                        kvp => kvp.Value.ToDictionary(
+                            kvp2 => kvp2.Key,
+                            kvp2 => kvp2.Value));
+
+                    int i = 0;
+                    foreach (var kvp1 in clothesData)
+                        foreach (var kvpKey in kvp1.Value.Keys.ToArray())
+                        {
+                            dict.Add(new KeyValuePair<int, ulong>(i, kvp1.Value[kvpKey].Hash), kvp1.Value[kvpKey].TextureBytes);
+                            kvp1.Value[kvpKey] = new ClothesTexData()
+                            {
+                                TextureBytes = BitConverter.GetBytes(i),
+                                BlendingMode = kvp1.Value[kvpKey].BlendingMode,
+                                Override = kvp1.Value[kvpKey].Override
+                            };
+                            i++;
+                        }
+                }
+                // KoiSkinOverlayX
+                else if (dictRaw is Dictionary<int, TextureHolder> skinData)
+                {
+                    foreach (var kvp in skinData)
+                        dict.Add(new KeyValuePair<int, ulong>(kvp.Key, kvp.Value.Hash), kvp.Value.Data);
+                }
+                else
+                {
+                    KoiSkinOverlayGui.Logger.LogError("Invalid data received for saving!");
+                    KoiSkinOverlayGui.Logger.LogMessage("[Overlays] Couldn't save overlays!");
+                    return;
+                }
+
+                var hashDict = dict.ToDictionary(pair => pair.Key.Key, pair => pair.Key.Value.ToString("X16"));
+                pluginData.data.Add(DedupedTexSavePrefix + DataKey, MessagePackSerializer.Serialize(hashDict));
+                if (clothesData != null)
+                    pluginData.data.Add(DedupedTexSavePrefix + key, MessagePackSerializer.Serialize(clothesData));
+
+                return;
+            }
+
+            // SceneController
+            var dicHashData = new Dictionary<ulong, byte[]>();
+            foreach (var controller in KoiSkinOverlayController.controllers)
+                foreach (var item in controller.OverlayStorage.TextureData.Where(x => !dicHashData.ContainsKey(x.Hash)))
+                    dicHashData.Add(item.Hash, item.Data);
+            foreach (var controller in KoiClothesOverlayController.controllers)
```

Comment:
Since controllers are used only here on a scene save I wouldn't bother keeping track of them. Just use FindObjectsOfType on both of the controller types and it will have basically no performance effect, and code will be cleaner.